tests/integration/saml_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package integration
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"code.gitea.io/gitea/models/auth"
+	"code.gitea.io/gitea/models/db"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/test"
+	"code.gitea.io/gitea/services/auth/source/saml"
+	"code.gitea.io/gitea/tests"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSAMLRegistration(t *testing.T) {
+	defer tests.PrepareTestEnv(t)()
+
+	samlURL := "localhost:8080"
+
+	if os.Getenv("CI") == "" || !setting.Database.Type.IsPostgreSQL() {
+		// Make it possible to run tests against a local simplesaml instance
+		samlURL = os.Getenv("TEST_SIMPLESAML_URL")
+		if samlURL == "" {
+			t.Skip("TEST_SIMPLESAML_URL not set and not running in CI")
+			return
+		}
+	}
+
+	privateKey, cert, err := saml.GenerateSAMLSPKeypair()
+	assert.NoError(t, err)
+
+	// verify that the keypair can be parsed
+	keyPair, err := tls.X509KeyPair([]byte(cert), []byte(privateKey))
+	assert.NoError(t, err)
+	keyPair.Leaf, err = x509.ParseCertificate(keyPair.Certificate[0])
+	assert.NoError(t, err)
+
+	assert.NoError(t, auth.CreateSource(db.DefaultContext, &auth.Source{
+		Type:          auth.SAML,
+		Name:          "test-sp",
+		IsActive:      true,
+		IsSyncEnabled: false,
+		Cfg: &saml.Source{
+			IdentityProviderMetadata:                 "",
+			IdentityProviderMetadataURL:              fmt.Sprintf("http://%s/simplesaml/saml2/idp/metadata.php", samlURL),
+			InsecureSkipAssertionSignatureValidation: false,
+			NameIDFormat:                             4,
+			ServiceProviderCertificate:               "", // SimpleSAMLPhp requires that the SP certificate be specified in the server configuration rather than SP metadata
```

Comment:
I don't think this is true? I've been using SimpleSAMLphp and it does accept SP certificates in the XML-format metadata directly.

Reply:
@grawity could you provide an example of your config for this so we can adjust the tests?

Reply:
I wasn't able to find the sources of the Docker image being used, so I don't know what it does with the SP metadata, but given the separate settings to specify ACS/SLO endpoints, I suspect it doesn't look at the XML metadata at all and just builds its own version [like this](https://github.com/druidfi/docker-images/blob/35c3ebb4638cadf5c00245bd58634e603aa1226e/misc/saml-idp/files/app/simplesamlphp/metadata/saml20-sp-remote.php)?

But on a full SimpleSAMLphp 2.x installation (and I believe 1.19.x as well), you can list the whole XML file in `config.php` and it will happily accept signed requests. (Although plain sign-on requests don't use the SP certificate at all; it's just a) encrypted assertions or b) signed single-logout requests that need it; but I re-tested the latter to verify that SimpleSAMLphp pays attention to the certificate in the SP's XML file and it does.)

```php
$config = [
    ...
    "metadata.sources" => [
        ["type" => "flatfile"],                             // the PHP-format saml20-sp-remote.php
        ["type" => "xml", "file" => "metadata/gitea.xml"],  // an XML-format metadata file
    ],
];
```

Even if you're using the PHP-format SP metadata file (`saml20-sp-remote.php`), where it might *look like* the certificate has to be in a separate file:

```php
$metadata["https://SOME_ENTITY_ID"] = [
    "AssertionConsumerService" => "https://SOME_URL",
    "certificate" => "exlibris-alma.crt",
    ...
];
```

it is still possible to specify it inline, and the XML-to-PHP metadata converter that comes with SimpleSAMLphp (Admin > Federation) will include any found certificates inline as part of the PHP-format metadata:

```php
$metadata["https://SOME_ENTITY_ID"] = [
    "AssertionConsumerService" => "https://SOME_URL",
    "keys" => [
        ["type" => "X509Certificate", "encryption" => true, "signing" => true, "X509Certificate" => "MII…"]
    ]
    ...
];
```

Reply:
Here's the repository with the configuration: https://github.com/AllSpiceIO/docker-test-saml-idp/tree/master

Reply:
I'm not certain on how thorough Gitea's requirements for test cases are, so take this with a grain of salt. But I don't think it's important to test SAML request signing because Gitea isn't making any claims that the IdP will rely on for the request. Also, (I sure hope) IdPs will be using HTTPS so there's no concern of tampering or confidentiality leaks anyways (not that there's any confidential data being sent by Gitea to begin with).